lto?: Wild runs out of file descriptors when building WebKit (`wild: error: Failed to open input file "..." Caused by: No file descriptors available (os error 24)`)

When compiling WebKitGTK with LTO and using wild as the linker, the final linking failed:
```bash
[8434/8453] : && /usr/lib/ccache/bin/x86_64-pc-linux-musl-clang++ \
-fPIC -fdiagnostics-color=always -fcolor-diagnostics -Wextra -Wall -Werror=undefined-internal -Werror=undefined-inline -pipe \
-Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-parentheses-equality -Qunused-arguments -Wundef -Wpointer-arith -Wmissing-format-attribute \
-Wformat-security -Wcast-align -Wno-tautological-compare -fasynchronous-unwind-tables -fdebug-types-section \
-march=native -O2 -pipe -flto=thin -fno-semantic-interposition -ftrivial-auto-var-init=zero -Werror=format-security -g1 \
-fno-strict-aliasing -fno-exceptions -fno-rtti -fcoroutines -ffunction-sections -fdata-sections  -shared -Wl,--no-undefined \
-Wl,-O1,--as-needed,--sort-common,--gc-sections,-Bsymbolic-functions,--build-id=sha1,-z,pack-relative-relocs,-z,relro,-z,now -flto -fuse-ld=wild \
-Wl,--gc-sections -Wl,--disable-new-dtags  -Wl,--version-script,/var/tmp/portage/net-libs/webkit-gtk-2.50.5-r600/work/webkitgtk-2.50.5/Source/WebKit/webkitglib-symbols.map \
-Wl,-soname,libwebkitgtk-6.0.so.4 -o lib/libwebkitgtk-6.0.so.4.13.8 @CMakeFiles/WebKit.rsp && :
ninja: job failed: : && /usr/lib/ccache/bin/x86_64-pc-linux-musl-clang++ \
-fPIC -fdiagnostics-color=always -fcolor-diagnostics -Wextra -Wall -Werror=undefined-internal -Werror=undefined-inline -pipe \
-Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-parentheses-equality -Qunused-arguments -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fasynchronous-unwind-tables -fdebug-types-section \
-march=native -O2 -pipe -flto=thin -fno-semantic-interposition -ftrivial-auto-var-init=zero -Werror=format-security -g1 \
-fno-strict-aliasing -fno-exceptions -fno-rtti -fcoroutines -ffunction-sections -fdata-sections  -shared -Wl,--no-undefined \
-Wl,-O1,--as-needed,--sort-common,--gc-sections,-Bsymbolic-functions,--build-id=sha1,-z,pack-relative-relocs,-z,relro,-z,now -flto -fuse-ld=wild \
-Wl,--gc-sections -Wl,--disable-new-dtags  -Wl,--version-script,/var/tmp/portage/net-libs/webkit-gtk-2.50.5-r600/work/webkitgtk-2.50.5/Source/WebKit/webkitglib-symbols.map \
-Wl,-soname,libwebkitgtk-6.0.so.4 -o lib/libwebkitgtk-6.0.so.4.13.8 @CMakeFiles/WebKit.rsp && :
WARNING: wild: -z pack-relative-relocs is not yet supported
wild: error: Failed to open input file `Source/WebCore/CMakeFiles/WebCore.dir/./page/scrolling/coordinated/ScrollingTreeFixedNodeCoordinated.cpp.o`
  Caused by:
    No file descriptors available (os error 24)

x86_64-pc-linux-musl-clang++: error: linker command failed with exit code 255 (use -v to see invocation)
ninja: subcommand failed
```
I thinks its a similar situation to the one with mold where it also ran out of file descriptors when using LTO with clang https://github.com/rui314/mold/issues/1362
I'll check later (probably this weekend) if this happens without LTO and/or with GCC but WebKit takes forever to compile (and loads of RAM) so will take a while to test for that.

Noticed the same issue when building the `ffmpeg` package on the Gentoo Linux:
```
wild: error: Failed to open input file `libavcodec/gif_parser.o`
  Caused by:
    Too many open files (os error 24)

collect2: error: ld returned 255 exit status
make: *** [/var/tmp/portage/media-video/ffmpeg-8.1/work/ffmpeg-8.1/ffbuild/library.mak:136: libavcodec/libavcodec.so.62] Error 1
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -61,6 +61,7 @@ libtest-mimic = "0.8.2"
 memchr = "2.7.0"
 memmap2 = "0.9.0"
 mimalloc = { version = "0.1", default-features = false }
+nix = { version = "0.30.1", default-features = false, features = ["resource"] }
 object = { version = "0.39.1", default-features = false, features = [
     # Note that compression is only actually needed for linker-diff
     "compression",

--- a/libwild/Cargo.toml
+++ b/libwild/Cargo.toml
@@ -36,6 +36,7 @@ libc = { workspace = true }
 libloading = { workspace = true, optional = true }
 memchr = { workspace = true }
 memmap2 = { workspace = true }
+nix = { workspace = true, optional = true }
 object = { workspace = true }
 perfetto-recorder = { workspace = true }
 rayon = { workspace = true }
@@ -67,7 +68,7 @@ tempfile = { workspace = true }
 fork = []
 
 # Support for linker-plugins.
-plugins = ["libloading"]
+plugins = ["libloading", "nix"]
 
 # Enable support for emitting a perfetto trace.
 perfetto = ["perfetto-recorder/enable", "tracing/release_max_level_info"]

--- a/libwild/src/linker_plugins.rs
+++ b/libwild/src/linker_plugins.rs
@@ -1241,11 +1241,30 @@ impl<'data> Store<'data> {
                 let Store::Loaded(loaded) = self else {
                     unreachable!();
                 };
+
+                // When the linker plugin is active, we keep LTO inputs open since the plugin API
+                // needs them. This can cause us to hit our file descriptor limit. To avoid this, we
+                // attempt to increase the limit. Increasing the file limit is best-effort. If we
+                // can't increase the file limit for some reason, continue without warning and hope
+                // we don't need too many open files.
+                let _ = increase_file_limit();
+
                 Ok(*loaded)
             }
             Store::Loaded(loaded_plugin) => Ok(*loaded_plugin),
         }
     }
+}
+
+/// Increase the soft file limit to whatever the hard limit is set to.
+fn increase_file_limit() -> Result {
+    use nix::sys::resource::Resource::RLIMIT_NOFILE;
+
+    let (_, hard_limit) = nix::sys::resource::getrlimit(RLIMIT_NOFILE)?;
+
+    nix::sys::resource::setrlimit(RLIMIT_NOFILE, hard_limit, hard_limit)?;
+
+    Ok(())
 }
 
 pub(crate) fn resolve_lto_symbols<'data, 'scope>(

--- a/wild/tests/integration_tests.rs
+++ b/wild/tests/integration_tests.rs
@@ -3296,7 +3296,7 @@ impl LinkCommand {
         }
 
         if self.linker.is_wild() && expectations.is_empty() && !output.is_empty() {
-            bail!("Unexpected output on {output_name}:\n{output}");
+            bail!("Unexpected output on {output_name}:\n{output}\nCommand:\n{self}");
         }
 
         Ok(())

